packages/core-contracts/contracts/utils/SetUtil.sol

```diff
@@ -0,0 +1,122 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+library SetUtil {
+    // ----------------------------------------
+    // Address support
+    // ----------------------------------------
+
+    struct AddressSet {
+        Bytes32Set raw;
+    }
+
+    function add(AddressSet storage set, address value) internal {
+        add(set.raw, bytes32(uint256(uint160(value))));
+    }
+
+    function remove(AddressSet storage set, address value) internal {
+        remove(set.raw, bytes32(uint256(uint160(value))));
+    }
+
+    function contains(AddressSet storage set, address value) internal view returns (bool) {
+        return contains(set.raw, bytes32(uint256(uint160(value))));
+    }
+
+    function length(AddressSet storage set) internal view returns (uint) {
+        return length(set.raw);
+    }
+
+    function valueAt(AddressSet storage set, uint position) internal view returns (address) {
+        return address(uint160(uint256(valueAt(set.raw, position))));
+    }
+
+    function positionOf(AddressSet storage set, address value) internal view returns (uint) {
+        return positionOf(set.raw, bytes32(uint256(uint160(value))));
+    }
+
+    function values(AddressSet storage set) internal view returns (address[] memory) {
```

Comment:
it can potentially revert for out of gas if not paginated (it's only addresses so the number should be really large, but possible)

Reply:
or maybe not... I'm not sure the cost of the assembly

Reply:
I think that we can ignore this concern. Please DM if you'd like to insist.